### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by nfurfaro**
Start refactoring Lock.sol and Ilock.sol.
Refactor modifiers in Lock.sol(this has broken 2 tests temporarily) and hard-coded `KeyreleaseMechanism` to "Public" for now, as this contract can only be public.
Rename ILock.sol to ILockPublic.sol for clarity.

## Note 
 I'm happy to do more work on this before it is merged, such as getting the 2 broken tests working, cleaning up my inline comments, etc...I just wanted to get the ball rolling with the PR process before going too far in any one direction, and get some feedback through "official chanels".

## Proposed Changes

  - Remove the concept of the `keyreleaseMechanism` from Lock contracts, as each type (public, private, restricted) can only ever be that type. This type selector would perhaps need to be moved to the Unlock contract, and Unlock.sol could be a dynamic factory contract.
  - After closer inspection, the erc721.org implementation of ERC721 seems more in line with our needs. several of its function definitions are already marked as `payable`(which is needed in Lock.sol), and the function visibilities tend to be `External`(prefered if the functions will not be called internally) as opposed to Public.

### CODE REVIEW

**Comment by julien51** [comment]
Hey, could you please use a better commit and PR message? This will be really hard to understand what this is about if we need to revert.

**Comment by julien51** [comment]
Also we cannot merge if tests are broken :(

**Comment by julien51** [comment]
Agreed about removing the `keyReleaseMechanism` which needs to be removed. I think it would be great to be able to eventually pass a smart contract name to Unlock's `createLock` function. 

As for ERC721, what do you suggest? Is it a problem that OZ's version is different from the version on erc721.org?

**Comment by nfurfaro** [comment]
I appreciate the feedback and should have indicated that this is a work in progress [WIP]. I'll clean this up. I'm guessing you'd prefer to deal with feedback or discussion in Slack prior to submitting a PR, and only submit when I think it's ready to be merged pending review.

File: smart-contracts/contracts/interfaces/IERC165.sol [hunk 1]
```diff
@@ -0,0 +1,19 @@
+pragma solidity 0.4.24;
+
+/**
+ * @title IERC165
+ * @dev https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md
+ */
+interface IERC165 {
```
[the remark this answers is not in the document]
Comment (nfurfaro) [reply smart-contracts/contracts/interfaces/IERC165.sol]: IERC165.sol is inherited by the open-Zepplin IERC721 interface. It is a part of the erc721.org implementation of the ERC721 interface, but had been commented out by (at)fulldecent.
It adds one function: 
`function supportsInterface(bytes4 interfaceID) external view returns (bool);`
which can be queried externally to see if a contract implements a given interface.
I don't know that we need this here. The compiler will ensure that our Lock contracts all implement the erc721 interface, this is for other contracts to call to determine if a Lock contract implements a given interface.

**Comment by nfurfaro** [comment]
> Agreed about removing the `keyReleaseMechanism` which needs to be removed. I think it would be great to be able to eventually pass a smart contract name to Unlock's `createLock` function.
> Let's discuss this on a call.
> As for ERC721, what do you suggest? Is it a problem that OZ's version is different from the version on erc721.org?
> It's not a problem that the 2 are slightly different, just a decision we need to make. 
If we use (a subset of)the OZ version, we either need to modify it :( or modify our functions in Lock.sol. 
If we use (a subset of)the erc721.org version, our functions can remain the same. I think the erc721.org version is more in line with what we need here.
> Please do not commit the built artifacts.
oops...I typically have /build in my .gitignore during development and didn't think about running 
`git add .` Will take care to not do this again.

File: package-lock.json [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
+    "version": "1.0.0",
+    "lockfileVersion": 1,
+    "requires": true,
+    "dependencies": {
+        "eslint-plugin-jest": {
+            "version": "21.22.1",
+            "resolved": "https://registry.npmjs.org/eslint-plugin-jest/-/eslint-plugin-jest-21.22.1.tgz",
+            "integrity": "sha512-OcVizyXljSps3nVfuPomfK8RKt7jvDsIsDrO7l1ZA4bDbiO9bEGWpsdU/x5F0pymVDG7ME88VlTboxbYJumLGQ=="
+        }
     }
```
[the remark this answers is not in the document]
Comment (nfurfaro) [reply package-lock.json]: Have switched my editor to use 2 spaces. Hopefully, that will fix the issue.

**Pull request closed without merging**